BJS2-26441-участвовать-в-событии

ifef228 opened the pull request.

ifef228 committed:
BJS2-26441-участвовать-в-событии

ifef228 committed:
форматирование

Lyroy-X commented on src/main/java/school/faang/user_service/controller/event/EventParticipationController.java:
```diff
@@ -0,0 +1,30 @@
+public class EventParticipationController {
+    private final EventParticipationService service;
+
+    public void registerParticipant(long eventId, UserDto userDto) {
```
Если не ошибаюсь лучше юзать Long 
![image](https://github.com/user-attachments/assets/8cfe77e5-28c1-4608-8aa3-fa2c2773a628)

ifef228 replied:
а зачем нам в метод null передавать? Мне кажется, что лучше на уровне компилятора запретить такие мувы, потому что это неправильные данные

Lyroy-X replied:
Но если у тебя придет 0 это тоже не правильные данные :3
Ну лан, давай лучше потом у Миши уточним. Мне нечем подкрепить свои слова

ifef228 replied:
валидатор на правильно введенные данные я сделал

miked4king replied:
стоит использовать long в случаях, когда ты на 100 процентов уверен что тебе не придет null

Okabe-Rintaro21 commented on src/main/java/school/faang/user_service/sevice/event/EventParticipationService.java:
```diff
@@ -0,0 +1,45 @@
+    private final UserMapper mapper;
+
+    public void registerParticipant(long eventId, long userId) {
+        if (repository.findAllParticipantsByEventId(eventId).stream()
+                .map(User::getId)
+                .anyMatch(id -> id == userId)) {
+            throw new IllegalArgumentException("cant register user: user already" + userId + " register to event " + eventId);
+        }
```
Лучше вынести в отдельный класс Validate в пакете validator, чтобы не писать if валидацию, а писать лишь метод.
![image](https://github.com/user-attachments/assets/09d54537-8dff-4c0a-baa1-8ebd5c2ecac3)
![image](https://github.com/user-attachments/assets/f966f7b4-ceb6-4420-b3e0-4893ba42ada8)
![image](https://github.com/user-attachments/assets/88db9248-d5d6-4a42-9b12-5fcf05605d66)

Lyroy-X replied:
Добавлю, что в дальнейшем, когда кода будет больше, будет тяжко разбираться в нем если мешать логику валидации и бизнес логику в одном месте. Так же тестирования в разы упрощается при таком подходе, хехе

Okabe-Rintaro21 commented on src/main/java/school/faang/user_service/sevice/event/EventParticipationService.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public void unregisterParticipant(long eventId, long userId) {
+        if (repository.findAllParticipantsByEventId(eventId).stream()
+                .map(User::getId)
+                .anyMatch(id -> id == userId)) {
+            repository.unregister(eventId, userId);
+        } else {
+            throw new IllegalArgumentException("cant unregister user: user do not" + userId + " register to event " + eventId);
+        }
+    }
```
Тоже нужен валидатор

Lyroy-X commented on src/main/java/school/faang/user_service/mapper/UserMapper.java:
```diff
@@ -0,0 +1,17 @@
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface UserMapper {
+    UserDto toDto(User user);
+
+    User toEntity(UserDto dto);
```
Нигде не юзаются :3

ifef228 replied:
метод toDto юзается бином, который спринг создает на основе этого интерфейса, для метода toDtoList, таким образом, логика этого метода становится чуть прощe. toEntity действительно не юзается, убрал

Lyroy-X replied:
хмм, если убрать toDto то mapper разберется сам, как замапить лист, ноуп?
Если я правильно понимаю метод нужен для большей читаемости? Если да, то чутка не согласен. Мне кажется и так ясно, что делает маппер

Lyroy-X commented on src/main/java/school/faang/user_service/sevice/event/EventParticipationService.java:
```diff
@@ -0,0 +1,45 @@
+@RequiredArgsConstructor
+public class EventParticipationService {
+    private final EventParticipationRepository repository;
+    private final UserMapper mapper;
```
Не юзается mapper

ifef228 replied:
в методе getParticipants этого класса юзается

Lyroy-X replied:
броу, я напутал чутка и поставил не туда коммент. Я хотел его оставить у мапера в тестах... 
Сделаю апрув, но лучше поправь это. Подмечу: в одном из тестов можно юзнуть маппер

ifef228 replied:
там если мапер не добавить, то в класса сервиса его значение будет null, я чисто для того, чтобы инжекция зависимостей правильно происходила, добавил

Murzin34 committed:
BJS2-26818 Организация события

ifef228 committed:
добавил класс EventParticipationServiceValidator.java, убрал неиспользуемые методы

ifef228 committed:
добавил тесты для валидатора, поменял тесты для сервиса

ifef228 committed:
поменял названия методов

Murzin34 committed:
Фикс.
Изменил реализацию фильтров.

Murzin34 committed:
Фикс.
Небольшой фикс EventFilterDto, EventTypeFilter, EventValidator.java. Поменял реализацию фильтров EventRelatedAllSkillsFilter и EventRelatedAnySkillsFilter на более простую.

Murzin34 committed:
Тесты.

ifef228 committed:
Merge remote-tracking branch 'origin/basilisk-master-stream6-BJS2-26441-участвовать-в-событии' into basilisk-master-stream6-BJS2-26441-участвовать-в-событии

# Conflicts:
#	src/main/java/school/faang/user_service/dto/UserDto.java
#	src/main/java/school/faang/user_service/mapper/UserMapper.java

ifef228 committed:
Merge remote-tracking branch 'origin/basilisk-master-stream6' into basilisk-master-stream6-BJS2-26441-участвовать-в-событии

# Conflicts:
#	src/main/java/school/faang/user_service/dto/UserDto.java
#	src/main/java/school/faang/user_service/mapper/UserMapper.java

ifef228 committed:
убрал чужие коммиты

ifef228 committed:
покрыл тестами контроллер и доубирал чужой код

miked4king commented on src/main/java/school/faang/user_service/controller/event/EventParticipationController.java:
```diff
@@ -0,0 +1,30 @@
+@Component
+@RequiredArgsConstructor
+public class EventParticipationController {
+    private final EventParticipationService service;
```
жестко завязываешься на конкретной реализации (нарушение последней буквы в solid)
хорошей практикой является использование интерфейсов, сделай EventParticipationService интерфейсом, сделай имплементацию, обычно называют название интерфейса + Impl (EventParticipationServiceImpl)

miked4king commented on src/main/java/school/faang/user_service/mapper/UserMapper.java:
```diff
@@ -5,8 +5,11 @@
 @Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
 public interface UserMapper {
-
     UserDto toDto(User user);
```
не вижу где используется

miked4king commented on src/main/java/school/faang/user_service/validator/EventParticipationServiceValidator.java:
```diff
@@ -0,0 +1,35 @@
+@Component
+@RequiredArgsConstructor
+public class EventParticipationServiceValidator {
+    private final EventParticipationRepository repository;
```
валидатор не должен ничего знать о репозитории, он должен проверять уже готовые данные

miked4king commented on src/main/java/school/faang/user_service/validator/EventParticipationServiceValidator.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public void validateUserUnregister(long eventId, long userId) throws IllegalArgumentException {
+        if (repository.findAllParticipantsByEventId(eventId).stream()
+                .map(User::getId)
+                .noneMatch(id -> id == userId)) {
+            throw new IllegalArgumentException("cant unregister user: user do not" + userId + " register to event " + eventId);
+        }
```
кажется дублирование кода с методом validateUserRegister.
может сделать так, чтобы метод возвращал boolean ?

miked4king commented on src/test/java/school/faang/user_service/controller/event/EventParticipationControllerTest.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @Test
+    void testRegisterParticipant() {
```
по названию методов в тестах, считаю что test в названии каждого метода излишне.
иногда названия становятся длинными и стоит разделять их на части через нижнее подчеркивание.
например registerParticipant_whenOk

ifef228 replied:
контроллер занимается чисто вызовом методов, поэтому я проверяю, что во всех методах все как надо вызвалось, и методы называю в честь оригиналов

miked4king replied:
хорошо, не в целом, когда ты пишешь ожидаемый результат становится легче ориентироваться в тестах.
они не всегда будут в пару строк

ifef228 replied:
принял

ifef228 committed:
добавил интерфейсы контроллера и сервиса, убрал репозиторий из валидатора, поправил тесты, поправил реализацию валидатора

miked4king commented on src/main/java/school/faang/user_service/controller/event/EventParticipationController.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.List;
+
+public interface EventParticipationController {
```
для контроллера не стоит создавать интерфейс, в большинстве случаев контроллеры представляют собой простые классы и создание интерфейса может только усложнить код.

miked4king commented on src/main/java/school/faang/user_service/validator/EventParticipationServiceValidator.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public boolean isUserRegistered(List<User> users,  long userId) {
+        if (users == null) {
```
по моему join (inner join) в случае, если у тебя нет совпадений вернет пустой список, тогда проверка на null будет избыточной

ifef228 committed:
поправил вопросы из ревью

miked4king commented:
разреши конфликты и можешь мержить, переведу в "готово"

ifef228 committed:
гитигнор

ifef228 committed:
гит игнор

ifef228 committed:
Merge branch 'basilisk-master-stream6' into basilisk-master-stream6-BJS2-26441-участвовать-в-событии

# Conflicts:
#	src/main/java/school/faang/user_service/dto/UserDto.java
#	src/main/java/school/faang/user_service/service/SubscriptionService.java
#	src/main/java/school/faang/user_service/service/SubscriptionServiceImpl.java
#	src/test/java/school/faang/user_service/service/SubscriptionServiceImplTest.java
